primitives/src/asset_identifier.rs

```diff
@@ -0,0 +1,207 @@
+use codec::{Decode, Encode};
+use core::convert::TryInto;
+use sp_std::vec;
+
+/// Implementation of common asset identifiers
+/// https://www.cusip.com/identifiers.html
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub enum AssetIdentifier {
+    /// Universally recognized identifier for financial instruments
+    /// Example: Amazon.com Inc - Common Stock
+    /// ISSUER ISSUE CHECK CUSIP
+    /// 023135 10    6     023135106
+    CUSIP([u8; 9]),
+    /// The CUSIP International Numbering System
+    /// Example: Abingdon Capital PLC - Shares
+    /// COUNTRY CODE ISSUER ISSUE CHECK CINS
+    /// G            0052B  10    5     G0052B105
+    CINS([u8; 9]),
+    /// The International Securities Identification Number
+    /// Example:
+    /// COUNTRY CODE LOCAL IDENTIFIER CHECK ISIN
+    /// CA           008911703        4     CA0089117034
+    ISIN([u8; 12]),
+    /// The Legal Entity Identifier
+    /// Example: Philadelphia Cheesesteak Company
+    /// LOU PREFIX ENTITY INDENTIFIER VERIFICATION ID LEI
+    /// 5493       00SAMIRN1R27UP     42              549300SAMIRN1R27UP42
+    LEI([u8; 20]),
+}
+
+impl AssetIdentifier {
+    /// Validate `bytes` is a valid CUSIP identifier, returns an instance of `Identifier` if successful
+    pub fn cusip(bytes: [u8; 9]) -> Option<AssetIdentifier> {
+        (cusip_checksum(&bytes[..8]) == bytes[8] - b'0').then_some(AssetIdentifier::CUSIP(bytes))
+    }
+
+    /// Validate `bytes` is a valid CINS identifier, returns an instance of `Identifier` if successful
+    pub fn cins(bytes: [u8; 9]) -> Option<AssetIdentifier> {
+        Self::cusip(bytes).map(|_| AssetIdentifier::CINS(bytes))
+    }
+
+    /// Validate `bytes` is a valid ISIN identifier, returns an instance of `Identifier` if successful
+    pub fn isin(bytes: [u8; 12]) -> Option<AssetIdentifier> {
+        let (s1, s2) = bytes
+            .iter()
+            .copied()
+            .map(byte_value)
+            .flat_map(|b| if b > 9 { vec![b / 10, b % 10] } else { vec![b] })
```

Comment:
Btw, if we wanted to gain in efficiency, we could use `ArrayVec` here instead to avoid the allocation, [or roll our own](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f80ccfc1b0af37e085fec0b645270ed4):
```rust
enum UpToTwo {
    Zero,
    One(u8),
    Two(u8, u8),
}
impl Iterator for UpToTwo {
    type Item = u8;
    fn next(&mut self) -> Option<Self::Item> {
        let (this, next) = match self {
            Self::Zero => return None,
            Self::One(x) => (Self::Zero, Some(*x)),
            Self::Two(x, y) => (Self::One(*y), Some(*x)),
        };
        *self = this;
        next
    }
}
```

and then:

```suggestion
            .flat_map(|b| if b > 9 { UpToTwo::Two(b / 10, b % 10) } else { UpToTwo::One(b) })
```